### ISSUE TITLE
Classcast error while using CosmosDB query with COUNT and VALUE

### ISSUE THREAD

**Issue by derryx** [Azure/azure-webjobs-sdk-extensions#569]
Hi!

I use Java to write a function that reads data from a CosmosDB instance. I encounter a problem when I use a query like `SELECT VALUE COUNT(1) FROM Messages m`.

I get a class cast exception:
System.Private.CoreLib: Exception while executing function: Functions.ReadDevicesLastSeenHistogram. Anonymously Hosted DynamicMethods Assembly: Unable to cast object of type 'Newtonsoft.Json.Linq.JValue' to type 'Newtonsoft.Json.Linq.JObject'.

I am using version 3.0.3 of the CosmosDB extension.

#### Repro steps
```
    @FunctionName("ReadDevicesLastSeenHistogram")
    public HttpResponseMessage readDevicesLastSeenHistogram(@HttpTrigger(name = "req",
            methods = {HttpMethod.GET},
            route = "statistics/lastSeenHistogram",
            authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<Void>> request,
            @CosmosDBInput(name="databaseAll",
                    databaseName = "IoTMessagesDatabase",
                    collectionName = "DevicesLastDocumentCollection",
                    connectionStringSetting = "siriusiotdb_DOCUMENTDB",
                    partitionKey = "/deviceId",
                    sqlQuery = "SELECT VALUE COUNT(1) FROM Messages m") String allCount,
            @CosmosDBInput(name="databaseMinus1",
                    databaseName = "IoTMessagesDatabase",
                    collectionName = "DevicesLastDocumentCollection",
                    connectionStringSetting = "siriusiotdb_DOCUMENTDB",
                    partitionKey = "/deviceId",
                    sqlQuery = "SELECT VALUE COUNT(1) FROM Messages m WHERE m._ts>udf.timestampMinusDays(1)") String dayCount){
        Map<String,Long> result=new HashMap<>();

        result.put("all",Long.valueOf(allCount));
        result.put("1day",Long.valueOf(dayCount));

        return request.createResponseBuilder(HttpStatus.OK).body(result).build();
    }
```
#### Expected behavior

No class cast error.

#### Actual behavior

System.Private.CoreLib: Exception while executing function: Functions.ReadDevicesLastSeenHistogram. Anonymously Hosted DynamicMethods Assembly: Unable to cast object of type 'Newtonsoft.Json.Linq.JValue' to type 'Newtonsoft.Json.Linq.JObject'

#### Known workarounds

None

#### Related information 

* Java Function App
* CosmosDB extension version 3.0.3

**Comment by joecoolish** [comment]
I am also getting this error.  I have the following `function.json` configuration for my Azure Function:

```
{
  "bindings": [
    {
      "name": "name",
      "type": "activityTrigger",
      "direction": "in"
    },
    {
      "type": "cosmosDB",
      "direction": "in",
      "name": "cableCount",
      "databaseName": "Cables",
      "collectionName": "CableData",
      "connectionStringSetting": "CosmosTrigger_ConnectionString",
      "sqlQuery": "SELECT VALUE COUNT(1) FROM c WHERE c.type='cable'"
    }
  ],
  "scriptFile": "../dist/GetCableCount/index.js"
}
```

and I get the following error message:

```
[3/25/2020 7:18:16 PM] System.Private.CoreLib: Exception while executing function: Functions.GetCableCount. Anonymously Hosted DynamicMethods Assembly: Unable to cast object of type 'Newtonsoft.Json.Linq.JValue' to type 'Newtonsoft.Json.Linq.JObject'.
[3/25/2020 7:18:16 PM] fb650787fc794364a8136b06400e7a22: Function 'GetCableCount (Activity)' failed with an error. Reason: System.InvalidCastException: Unable to cast object of type 'Newtonsoft.Json.Linq.JValue' to type 'Newtonsoft.Json.Linq.JObject'.
[3/25/2020 7:18:16 PM]    at CallSite.Target(Closure , CallSite , Object )
[3/25/2020 7:18:16 PM]    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.Documents.Client.FeedResponseBinder.Convert[T](FeedResponse`1 dynamicFeed)
[3/25/2020 7:18:16 PM]    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.Documents.Linq.DocumentQuery`1.ExecuteNextPrivateAsync[TResponse](CancellationToken cancellationToken)
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.WebJobs.Extensions.CosmosDB.CosmosDBService.ExecuteNextAsync[T](Uri documentCollectionUri, SqlQuerySpec sqlSpec, String continuation)
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.WebJobs.Extensions.CosmosDB.CosmosDBEnumerableBuilder`1.ConvertAsync(CosmosDBAttribute attribute, CancellationToken cancellationToken)
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.WebJobs.Extensions.CosmosDB.Bindings.CosmosDBJArrayBuilder.ConvertAsync(CosmosDBAttribute attribute, CancellationToken cancellationToken) in C:\azure-webjobs-sdk-extensions\src\WebJobs.Extensions.CosmosDB\Bindings\CosmosDBJArrayBuilder.cs:line 22
[3/25/2020 7:18:16 PM]    at Microsoft.Azure.WebJobs.Host.Bindings.PatternMatcher.AsyncInvoker`2.<>c__DisplayClass0_0.<<Execute>b__0>d.MoveNext()
```

Is there any way to get the count?

**Comment by v-anvari** [comment]
Hi @derryx , Were you about to find a solution here, please let us know if you have are facing the same issues so that we can look into it further.

**Comment by derryx** [comment]
I never have been able to find a solution to this. I switched to an alternate way to get the information I need out of CosmosDB.

But all Information to reproduce the issue are in the description of the ticket so I am not sure what you need from me.

**Comment by v-anvari** [comment]
Hi @ealsur / @brettsam , Can you please go through the error and let us know if the user has missed any addition which resulted in the error

**Comment by joergkrause** [comment]
I can confirm that this even happens in a JavaScript/nodejs function, so it's entirely on the input binding side. The error is, however, like this:

**System.Private.CoreLib: Exception while executing function: Functions.statistics. Newtonsoft.Json: Can not convert from System.Int64 to Newtonsoft.Json.Linq.JObject.**

You can modify the query like this:  

`"sqlQuery": "SELECT COUNT(1) FROM c WHERE c.type='cable'"`

This returns the full result, which requires additional work in the function code.

`
[
 {
   "$1": 1000 
}
]
`

A solution that accepts scalar values would be great indeed.

**Comment by ealsur** [comment]
The problem is related to the fact that languages other than C# invoke the binding with a JArray C# type. In C# it would be like defining a trigger like:

```
[CosmosDB("myDb", "myCollection", SqlQuery = "SELECT VALUE COUNT(1) FROM c")] JArray jArray
```

JArray is an array of JObjects where each one is a Json document.

But the query in itself does not yield Json documents, it yields a list of numbers.

In C#, the correct way to execute such a query is:

```
[CosmosDB("test", "test", SqlQuery = "SELECT VALUE COUNT(1) FROM c")] IEnumerable<JValue> results
```

The question here is, from a design perspective, I don't think there is a way for the other languages to do any changes in the way the underlying .NET layer is invoked, it will always use JArray. If there was such a way, then the binding would work. The problem is not the binding in itself, but rather the assumption of the language translation layer that assumes the result will always be JArray.

**Comment by ealsur** [comment]
There might be a change that can be done in this file: https://github.com/Azure/azure-webjobs-sdk-extensions/blob/dev/src/WebJobs.Extensions.CosmosDB/Bindings/CosmosDBJArrayBuilder.cs#L13-L24

And instead of JObject use JToken, which might work, because JValue and JObject are both JToken.

**Comment by ealsur** [comment]
Applying that change makes the Function work and execute, but the output would be a String that has the JArray with the number, for example:

```
[ 10 ]
```

It will always be an array, because the translation layer uses JArray.

**Comment by joergkrause** [comment]
@ealsur That would be perfect, because it's exactly the same result when you use the Data Explorer in the portal. That way you can check the query first for proper execution and get the same result in the function. And, taking the first element of an array is easy enough in any of the supported languages.

**Comment by ealsur** [comment]
I can try to send a PR this week with this requested change.

**Comment by ealsur** [comment]
The fix will be available in extension 3.0.10 coming in the next week.
